src/rgw/rgw_sal_d3n.cc

```diff
@@ -0,0 +1,307 @@
+      id += diff_ofs;
+    }
+    uint64_t read_ofs = diff_ofs; //read_ofs is the actual offset to start reading from the current part/ chunk
+    std::string oid_in_cache = source->get_bucket()->get_marker() + "_" + oid + "_" + std::to_string(adjusted_start_ofs) + "_" + std::to_string(part_len);
+    rgw_raw_obj r_obj;
+    r_obj.oid = oid_in_cache;
+    ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << __func__ << "(): READ FROM CACHE: oid=" << oid_in_cache << " length to read is: " << len_to_read << " part num: " << start_part_num << " read_ofs: " << read_ofs << " part len: " << part_len << dendl;
+    if (filter->get_d3n_cache()->get(oid_in_cache, part_len)) {
+      // Read From Cache
+      auto completed = aio->get(r_obj, rgw::Aio::d3n_cache_op(dpp, y, read_ofs, len_to_read, filter->get_d3n_cache()->cache_location), cost, id);
+      ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << __func__ << "(): Info: flushing data for oid: " << oid_in_cache << dendl;
+      auto r = flush(dpp, std::move(completed));
+      if (r < 0) {
+        ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << __func__ << "(): Error: failed to flush, r= " << r << dendl;
+        return r;
+      }
+    } else {
+      //for ranged requests, for last part, the whole part might exist in the cache
+      oid_in_cache = source->get_bucket()->get_marker() + "_" + oid + "_" + std::to_string(adjusted_start_ofs) + "_" + std::to_string(obj_max_req_size);
+      r_obj.oid = oid_in_cache;
+      ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << __func__ << "(): READ FROM CACHE: oid=" << oid_in_cache << " length to read is: " << len_to_read << " part num: " << start_part_num << " read_ofs: " << read_ofs << " part len: " << part_len << dendl;
+      if (filter->get_d3n_cache()->get(oid_in_cache, obj_max_req_size)) {
+        // Read From Cache
+        auto completed = aio->get(r_obj, rgw::Aio::d3n_cache_op(dpp, y, read_ofs, len_to_read, filter->get_d3n_cache()->cache_location), cost, id);
+        ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << __func__ << "(): Info: flushing data for oid: " << oid_in_cache << dendl;
+        auto r = flush(dpp, std::move(completed));
+        if (r < 0) {
+          ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << __func__ << "(): Error: failed to flush, r= " << r << dendl;
+          return r;
+        }
+      } else {
+        ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << __func__ << "(): Info: draining data for oid: " << oid_in_cache << dendl;
+        auto r = drain(dpp);
+        if (r < 0) {
+          ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << __func__ << "(): Error: failed to drain, r= " << r << dendl;
+          return r;
+        }
+        break;
+      }
+    }
+    if (start_part_num == (num_parts - 1)) {
+      ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << __func__ << "(): Info: draining datafor oid: " << oid_in_cache << dendl;
+      return drain(dpp);
```

Comment:
@dang : the method returns from here, if all parts are successfully fetched from the cache, else once a missing part is encountered, the missing part and the subsequent parts are read from the backend store.

Reply:
Ah thanks, I missed that.